clippy_lints/src/replace_box.rs

```diff
@@ -109,3 +148,27 @@ fn get_box_new_payload<'tcx>(cx: &LateContext<'_>, expr: &Expr<'tcx>) -> Option<
         None
     }
 }
+
+struct MovedVariablesCtxt<'a, 'tcx> {
+    cx: &'a LateContext<'tcx>,
+    moved_vars: HirIdSet,
+}
+
+impl<'tcx> Delegate<'tcx> for MovedVariablesCtxt<'_, 'tcx> {
+    fn consume(&mut self, cmt: &PlaceWithHirId<'tcx>, _: HirId) {
+        if let PlaceBase::Local(vid) = cmt.place.base
+            && let Node::Expr(expr) = self.cx.tcx.hir_node(cmt.hir_id)
+            && expr.res_local_id().is_some_and(|hir_id| hir_id == vid)
```

Comment:
This doesn't work if the box is a field.

Reply:
Right.

What do you think of https://github.com/samueltardieu/rust-clippy/commit/f767523f73f6?

@profetia Feel free to cherry-pick and squash it

Reply:
That won't work with nested boxes. e.g.

```rust
let x = Box::new(Box::new(String::new()));
*x;
*x = Box::new(String::new());
```

---

I don't even know why you're looking at the place expression in the first place. Since you want to make an exception for the deref projection you should be checking for that. e.g.

```rust
let mut projections = cmt.place.projection.iter().filter(|x| matches!(x.kind, ProjectionKind::Deref));
if projections.next().is_none() || projections.next().is_some() {
  // Either no deref or multiple derefs
}
```

Linting on the MIR would be the correct thing to do since you want liveness analysis, but this should at least remove the false positives.

Reply:
The full visitor function should be:
```rust
fn consume(&mut self, cmt: &PlaceWithHirId<'tcx>, _: HirId) {
    if let PlaceBase::Local(id) = cmt.place.base
        && let mut projections = cmt
            .place
            .projections
            .iter()
            .filter(|x| matches!(x.kind, ProjectionKind::Deref))
        && (projections.next().is_none() || projections.next().is_some())
    {
        self.moved_vars.insert(id);
    }
}
```